### DISCUSSION TITLE
How to view pmtiles locally?

### DISCUSSION THREAD

**Discussion by Robinlovelace** [protomaps/PMTiles#234] [General]
I've got a local test version of a pmtiles file and am trying to view it in our statically hosted web app on localhost.

This is the change I've made:

![image](https://github.com/protomaps/PMTiles/assets/1825120/357f76c5-3621-4e4d-8996-b53f91ac8162)

But the tiles are not showing. Any advice greatly appreciated, cc @mem48 who wrote most of the .js code.

Source: https://github.com/nptscot/nptscot.github.io/pull/76/files

**Comment by Robinlovelace** [comment]
Solution: 

```
    	'url': 'pmtiles://rnet_2023-08-11.pmtiles',

```

**Reply by bdon** [reply]
That looks correct if `rnett_2023-08-11.pmtiles` lives next to the page that's loading the style, but may not work if you want to load the style on a page with a different path... 

An absolute path that starts with `/` may work in that case

**Comment by jaguardo** [comment]
I've tried this many different ways for locally stored pmtile files (using jsx/react) and keep getting a `Error: Wrong magic number for PMTiles archive` is there something subtle that could be missing?

**Reply by jaguardo** [reply]
I also created a sandbox to demonstrate:
[sandbox link](https://codesandbox.io/p/sandbox/nice-scooby-sm9hz9?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clvpn93gf0006356ip9ajnk06%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clvpn93gf0002356ig6y2v7li%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clvpn93gf0003356ir0e53wem%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clvpn93gf0005356ipugt00nf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B53.86505592197983%252C46.13494407802017%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clvpn93gf0002356ig6y2v7li%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clvpn93ge0001356iyhk33vyh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clvpn93gf0002356ig6y2v7li%2522%252C%2522activeTabId%2522%253A%2522clvpn93ge0001356iyhk33vyh%2522%257D%252C%2522clvpn93gf0005356ipugt00nf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clvpn93gf0004356i2tozb0yh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clvpn93gf0005356ipugt00nf%2522%252C%2522activeTabId%2522%253A%2522clvpn93gf0004356i2tozb0yh%2522%257D%252C%2522clvpn93gf0003356ir0e53wem%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clvpn93gf0003356ir0e53wem%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

if the remote version of the pmtile is used:
```
      mapStyle={{
        version: 8,
        sources: {
          sample: {
            type: "vector",
            url:
              "pmtiles://https://r2-public.protomaps.com/protomaps-sample-datasets/cb_2018_us_zcta510_500k.pmtiles"
              // "pmtiles://cb_2018_us_zcta510_500k.pmtiles"
            }
        },
```

the map shows up
![image](https://github.com/protomaps/PMTiles/assets/28798269/9f10aad8-796b-447f-9449-89abcbb753ea)

if the remote location is commented out, and the local address used (pmtile was copied to the sandbox)
![image](https://github.com/protomaps/PMTiles/assets/28798269/f971583a-576f-4bb0-be2d-f3115c74039f)


```
            url:
              // "pmtiles://https://r2-public.protomaps.com/protomaps-sample-datasets/cb_2018_us_zcta510_500k.pmtiles"
              "pmtiles://cb_2018_us_zcta510_500k.pmtiles"
            }
```

the map does not show up, and this error appears:
```
proxyConsole.js:64 Error: Server returned no content-length header or content-length exceeding request. Check that your storage backend supports HTTP Byte Serving.
    at FetchSource.eval (index.js:1079:15)
    at Generator.next (<anonymous>)
    at fulfilled (index.js:123:24)
```
Does this imply that the file has to be "served" to be accessible? would there be a way to access it from the local folder structure?

**Reply by bdon** [reply]
It depends on what build system or local HTTP server you're using, you should put it in the `public/` folder like any other static assets.

**Reply by jaguardo** [reply]
I tried putting it in the public folder, in fact i think i updated the sandbox. It is just a react dev atm using 'yarn start' but it is pretty repeatable so far... but yarn/react doesn't serve files like a traditional http server.

**Reply by jaguardo** [reply]
![image](https://github.com/protomaps/PMTiles/assets/28798269/ffd01f52-05b6-491e-89a8-867c5593fb01)
![image](https://github.com/protomaps/PMTiles/assets/28798269/7679387c-6f49-43bd-afa7-b2169bcb375d)

**Reply by bdon** [reply]
your issue is specific to code sandbox, please verify that things uploaded to CSB public/ support HTTP 1.1 byte serving

**Reply by jaguardo** [reply]
Interesting... it suddenly started working on codesandbox this am (at least it seems hit or miss at times)... dont think I changed anything, so maybe the internal process that codesandbox uses for uploads took awhile to work through its process.

I did have to regress to earlier versions of some packages (maplibre-gl and react-map-gl) to get it to work locally, but have not trouble shot that any further.

**Comment by giswqs** [comment]
Here is a Python example for visualizing local PMTiles. 
https://leafmap.org/notebooks/93_maplibre_pmtiles/#local-pmtiles
